msger_mosaic_bridge/src/mosaic_adapter.cpp

```diff
@@ -0,0 +1,271 @@
+    RCLCPP_INFO(this->get_logger(), 
+                "Received Pose - Lat: %f, Lon: %f, Alt: %f", 
+                pose[0], pose[1], pose[2]);
+
+    gps_msgs::msg::GPSFix gps_msg;
+    gps_msg.latitude = pose[0];  
+    gps_msg.longitude = pose[1]; 
+    gps_msg.altitude = pose[2];  
+
+    RCLCPP_INFO(this->get_logger(), 
+                "GPSFix Message - Lat: %f, Lon: %f, Alt: %f", 
+                gps_msg.latitude, gps_msg.longitude, gps_msg.altitude);
+
+    gps_pub_->publish(gps_msg);
+}
+
+
+void MosaicAdapter::on_vehicle_twist_received_handler(const std::array<double, 3>& twist) {
+    RCLCPP_INFO(this->get_logger(), 
+            "Received Twist - x: %f, y: %f, z: %f", 
+            twist[0], twist[1], twist[2]);
+
+    geometry_msgs::msg::TwistStamped twist_msg;
+    twist_msg.twist.linear.x = twist[0];
+    twist_msg.twist.linear.y = twist[1];
+    twist_msg.twist.linear.z = twist[2]; 
+
+    twist_msg.header.stamp = this->now();
+
+    RCLCPP_INFO(this->get_logger(), 
+                "TwistStamped Message - Linear: x=%f, y=%f, z=%f", 
+                twist_msg.twist.linear.x, twist_msg.twist.linear.y, twist_msg.twist.linear.z);
+
+    twist_pub_->publish(twist_msg);
+}
+
+
+void MosaicAdapter::on_time_received_handler(unsigned long timestamp)
+{
+    rosgraph_msgs::msg::Clock time_now;
+
+    auto chrono_time = std::chrono::system_clock::now();
+    auto epoch = chrono_time.time_since_epoch();
+    auto milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(epoch);
+    auto time_to_milli = static_cast<int>(timestamp / 1e6);
+
+    RCLCPP_DEBUG(rclcpp::get_logger("MosaicClient"), 
+                 "Simulation Time: %d where current system time is: %ld", 
+                 time_to_milli, milliseconds.count());
+
+    time_now.clock.sec = static_cast<int>(timestamp / 1e9);
+    time_now.clock.nanosec = static_cast<uint32_t>(timestamp - time_now.clock.sec * 1e9);
+
+    time_pub_->publish(time_now);
+}
+
+void MosaicAdapter::on_siren_and_light_status_recieved_handler(bool siren_active, bool light_active){
+    uint8_t status_code = 0;
+    if (!siren_active && !light_active) {
+        status_code = SIRENS_AND_LIGHTS_INACTIVE;
+    } else if (siren_active && !light_active) {
+        status_code = ONLY_SIRENS_ACTIVE;
+    } else if (!siren_active && light_active) {
+        status_code = ONLY_LIGHTS_ACTIVE;
+    } else if (siren_active && light_active) {
+        status_code = SIRENS_AND_LIGHTS_ACTIVE;
+    }
+    
+    auto message = std::make_shared<std::vector<uint8_t>>(1, status_code);
+    mosaic_client_.send_siren_and_light_message(message);
+}
+
+
+std::string MosaicAdapter::compose_handshake_msg(const std::string& role_id, 
+                                                 int veh_status_port, 
+                                                 int time_port, 
+                                                 const std::string& ip)
+{
+    // document is the root of a json message
```

Comment:
Something weird with the indentation here, wouldn't hurt to run a formatter on this or just clean it up by hand.

Reply:
I didn't notice any unusual indentation on my end, but I've re-run the formatting just to be sure.